File: scripts/glue/ad_clicks_etl.scala
import com.amazonaws.services.glue.ChoiceOption
import com.amazonaws.services.glue.GlueContext
import com.amazonaws.services.glue.MappingSpec
import com.amazonaws.services.glue.ResolveSpec
import com.amazonaws.services.glue.errors.CallSite
import com.amazonaws.services.glue.util.GlueArgParser
import com.amazonaws.services.glue.util.Job
import com.amazonaws.services.glue.util.JsonOptions
import org.apache.spark.SparkContext
import scala.collection.JavaConverters._

object GlueApp {
  val argNames = Seq(
    "JOB_NAME",
    "database-name",
    "raw-table-name",
    "table-name"
  )

  val mappings = Seq(
    MappingSpec("user", "string", "user", "string"),
    MappingSpec("ad", "string", "ad", "string"),
    MappingSpec("at", "timestamp", "at", "timestamp"),
    MappingSpec("partition_0", "int", "year", "int"),
    MappingSpec("partition_1", "int", "month", "int")
  )

  def main(sysArgs: Array[String]) {
    val spark: SparkContext = new SparkContext()
    val glueContext: GlueContext = new GlueContext(spark)
    val args = GlueArgParser.getResolvedOptions(sysArgs, argNames.toArray)
    Job.init(args("JOB_NAME"), glueContext, args.asJava)

    // val databaseName = args("database-name")
    // val rawTableName = args("raw-table-name")
    // val tableName = args("table-name")

    val adClicksPreCast = glueContext.getCatalogSource("warehouse", "raw_ad_clicks").
      getDynamicFrame()

    adClicksPreCast.printSchema()
    adClicksPreCast.toDF().show()

    val adClicksPreMapping = adClicksPreCast.
      resolveChoice(Seq(
        ("at", "cast:timestamp"),
        ("partition_0", "cast:int"),
        ("partition_0", "cast:int")))

    adClicksPreMapping.printSchema()
    adClicksPreMapping.toDF().show()

    val adClicksPreResolve = adClicksPreMapping.
      applyMapping(mappings, false)

    adClicksPreResolve.printSchema()
    adClicksPreResolve.toDF().show()

    val adClicks = adClicksPreResolve.
      resolveChoice(Seq.empty[ResolveSpec], Some(ChoiceOption("MATCH_CATALOG")), Some("warehouse"), Some("ad_clicks"))

    adClicks.printSchema()
    adClicks.toDF().show()

    val partitioning = JsonOptions("""{"partitionKeys":["year","month"]}""")
    glueContext.getCatalogSink("warehouse", "ad_clicks", additionalOptions = partitioning).
      writeDynamicFrame(adClicks)

    Job.commit()
  }
}

File: scripts/glue/ad_clicks_etl.py
import sys
from awsglue.transforms import *
from awsglue.utils import getResolvedOptions
from pyspark.context import SparkContext
from awsglue.context import GlueContext
from awsglue.job import Job

sc = SparkContext()
glueContext = GlueContext(sc)

job = Job(glueContext)
args = getResolvedOptions(sys.argv, ['JOB_NAME', "database_name", "raw_table_name", "table_name"])
job.init(args['JOB_NAME'], args)

database_name = args["database_name"]
raw_table_name = args["raw_table_name"]
table_name = args["table_name"]

mappings = [("at",          "string", "at",    "timestamp"),
            ("user",        "string", "user",  "string"),
            ("ad",          "string", "ad",    "string"),
            ("partition_0", "string", "year",  "int"),
            ("partition_1", "string", "month", "int")]

ad_clicks = glueContext.create_dynamic_frame_from_catalog(database_name, raw_table_name) \
            .apply_mapping(mappings) \
            .resolveChoice(choice = "MATCH_CATALOG", database = database_name, table_name = table_name)

glueContext.write_dynamic_frame.from_catalog(ad_clicks,
                                             database_name,
                                             table_name,
                                             additional_options = {"partitionKeys": ["year", "month"]})

job.commit()
